src/analyses/apron/relationAnalysis.apron.ml

```diff
@@ -296,12 +293,17 @@ struct
     let new_rel = RD.add_vars st.rel arg_vars in
     (* RD.assign_exp_parallel_with new_rel arg_assigns; (* doesn't need to be parallel since exps aren't arg vars directly *) *)
     (* TODO: parallel version of assign_from_globals_wrapper? *)
-    let ask = Analyses.ask_of_ctx ctx in
-    let new_rel = List.fold_left (fun new_rel (var, e) ->
-        assign_from_globals_wrapper ask ctx.global {st with rel = new_rel} e (fun rel' e' ->
-            RD.assign_exp rel' var e' (no_overflow ask e)
-          )
-      ) new_rel arg_assigns
+    let new_rel =
+      if thread then
+        (* TODO: Why does test 63/16 not reach fixpoint without copy here? *)
+        RD.copy new_rel
```

Comment:
That is indeed worth investigating, any mutable state should be hidden!

Reply:
This `RD.copy` was necessary due to the destructive `RD.remove_filter_with` happening later. I now moved the `RD.copy` up, to ensure that it is done before the `RD.remove_filter_with`.

Reply:
`RD.add_vars` isn't a `_with` function, so it should already do a copy. Unless there's something wrong there.

Reply:
The `add_vars` of the `AffineEqualityDomain` does not do any explicit copying. It calls `SharedFunctions.EnvOps.add_vars`, which is defined as follows:
https://github.com/goblint/analyzer/blob/a651cccf69af65581222d7e8e046b8ad67510cf4/src/cdomains/apron/sharedFunctions.apron.ml#L303-L310

I assume that in case the list of vars that is added is empty, the `add_vars` for affine equalities will return the same physical object as its argument.

Reply:
In that case I think such optimization in `AffineEqualityDomain` is incorrect because as far as I am aware, all non-`_with` functions should provide a pure functional interface with no unintended sharing between the inputs and outputs. That's what makes it possible to write any imperative code with `_with` functions without having to worry about that modifying something which isn't a direct argument of the imperative function.

Reply:
I added an explicit copy operation to the `add_vars`, `drop_vars`, `keep_vars` and `keep_filter`  operations in `AffineEqualityDomain`. The `copy` in `RelationAnalysis.make_calleee_rel` thus was removed.